lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+    }
+    token_.set_kind(TokenKind::None);
+    break;
+  }
+
+  token_.set_length(cursor_ - format_ - token_.offset() + 1);
+}
+
+template<typename CHAR> int FormatValidator<CHAR>::Check() {
+  if (format_ == nullptr || !*format_) {
+    AppendError("Empty format expression");
+    return errorCount_;
+  }
+  NextToken();
+  if (token_.kind() != TokenKind::LParen) {
+    AppendError("Format expression must have an initial '('");
+    return errorCount_;
+  }
+  NextToken();
+
+  int nestLevel{0};  // Outer level ()s are at level 0.
+  Token starToken{};  // unlimited format token
+  bool hasDataEditDesc{false};
+
+  // Subject to error processing exceptions, a loop iteration processes an
+  // edit descriptor or does list management.  The loop terminates when
+  //  - a level-0 right paren is processed
+  //  - the end of an incomplete format is reached
+  //  - a threshold caller-chosen number of errors have been diagnosed
+  while (errorCount_ < maxErrors_) {
+    Token signToken{};
+    Token knrToken{};  // (nonnegative) k, n, or r value
+    int64_t knrValue{-1};  // -1 ==> not present
+    int64_t wValue{-1};
+    bool commaRequired{true};
+
+    if (token_.kind() == TokenKind::Sign) {
+      signToken = token_;
+      NextToken();
+    }
+    if (token_.kind() == TokenKind::UnsignedInteger) {
+      knrToken = token_;
+      knrValue = integerValue_;
+      NextToken();
+    }
+    if (signToken.IsSet() && (knrValue < 0 || token_.kind() != TokenKind::P)) {
+      argString_[0] = format_[signToken.offset()];
+      argString_[1] = 0;
+      AppendError(
+          "Unexpected '%s' in format expression", signToken, argString_);
+    }
+    // Default error argument.
+    // Alphabetic descriptor names are one or two characters in length.
+    argString_[0] = toupper(format_[token_.offset()]);
+    argString_[1] = token_.length() > 1 ? toupper(*cursor_) : 0;
+
+    auto check_r = [&](bool allowed = true, const char *name = {}) -> void {
+      if (!allowed && knrValue >= 0) {
+        AppendError("Repeat specifier before %s descriptor", knrToken, name);
+      } else if (knrValue == 0) {
+        AppendError(  // C1304
+            "%s descriptor repeat specifier must be positive", knrToken, name);
+      }
+    };
+
+    // Return the predicate "w value is present" to control further processing.
+    [[maybe_unused]] auto check_w = [&](bool required = true) -> bool {
+      if (token_.kind() == TokenKind::UnsignedInteger) {
+        wValue = integerValue_;
+        if (wValue == 0 &&
+            (*argString_ == 'A' || *argString_ == 'L' ||
+                stmt_ == IoStmtKind::Read)) {  // C1306, 13.7.2.1p6
+          AppendError("%s descriptor 'w' value must be positive");
+        }
+        NextToken();
+        return true;
+      }
+      if (required) {
+        AppendError("Expected %s descriptor 'w' value");  // C1306
+      }
+      return false;
+    };
+
+    auto check_m = [&]() -> void {
```

Comment:
Why are these lambdas rather than member functions? It obscures the fact that local variables are being accessed and even modified.

Reply:
There are trade-offs in doing this, but fewer drawbacks than I expected.  Done.